smartcontractkit/external-adapters-js
# Fallback to local cache when there are issues connecting to the Redis cache

## Closes 23191

## Description

This ticket implements the fallback method for adapters to start using their local in memory cache if the connection to Redis fails.

## Changes

- Implement fallback method to handle Redis cache failures
- Make local LRU cache singleton



## Steps to Test

Testing single requests

1. Run `docker compose up` from `/grafana` folder to start Redis cluster
2. Send requests to ensure that the adapter is caching requests correctly using Redis. 
3. Shutdown Redis
4. Send requests and ensure that requests were still being cached.

Testing WS requests

1. Enable websockets by running `export WS_ENABLED=true`
2. Follow the instructions to test single requests
3. Verify that the adapter saves WS results by searching for this log `WS: Saved result`.  Note the result.
4. Send another request to the adapter and verify that the value matches what is saved and that the response came from the cache.

Note:  You can tell that requests are being cached by turning setting `API_VERBOSE` to true and verifying that the `cacheHit` flag in the response has been set to true.

Example

```
{
    "jobRunID": "278c97ffadb54a5bbb93cfec5f7b5503",
    "result": 4310.24,
    "maxAge": 120000,
    "debug": {
        "cacheHit": true,
        "staleness": 0.789,
        "performance": 0.003814976,
        "providerCost": 0
    },
    "statusCode": 200,
    "data": {
        "result": 4310.24
    }
}
```

## Quality Assurance
